### ISSUE TITLE
Cannot bind to any IPv6 addresses

### ISSUE THREAD

**Issue by mynameisneo7** [getdnsapi/stubby#302]
- Ubuntu 21.10
- Raspberry Pi 4 Model B

I've looked through a few bugs reported previously with issues binding to IPv6 addresses (e.g. #261 and #281), with some workarounds but not quite like the issue I'm currently having. Even the default config address of `0::1` does not work; only IPv4 addresses are working for me.

Trying to track the issue down, I receive lots of very strange messages like "Cannot assign requested address," "File exists," and "Success." This is despite verifying that the ports and addresses are not in use-- my tests were also done as root.

Using strace to log what the app is actually doing gives very strange behavior; for the "File exists" and "Success" messages, stubby is not even bothering to attempt to bind any address, it is simply making `$HOME/.getdns` and returning that as the result. This typically occurred when I used a valid IPv6 address on eth0.

When using "::" or "::0" or "0::0" however, the error "Cannot assign requested address" is usually given. Again, checking strace reveals curious logs like this:

> [ 294] setsockopt(3, SOL_TCP, TCP_FASTOPEN, [1], 4) = -1 ENOPROTOOPT (Protocol not available)
> [ 282] bind(3, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("48.58.58.48")}, 16) = -1 EADDRNOTAVAIL (Cannot assign requested address)
> [   6] close(3)                         = 0

Note that the address `48.58.58.48` is a perfect ASCII representation of the string `0::0`, so it seems as though stubby is simply converting the text to binary and using it for the address to bind to, it doesn't appear to detect that it is an IPv6 address at all. I also tried `0::1` out of curiosity and the bind attempt shows `48.58.58.49` following the same pattern.

**Comment by saradickinson** [comment]
Thanks for the report - this does indeed sound odd. I only have a 20.04 to hand right now where this works and can confirm that what you should be seeing using `strace` is
`bind(3, {sa_family=AF_INET6, sin6_port=htons(53), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "::1", &sin6_addr), sin6_scope_id=0}, 28) = 0`
So the parsing of the address is incorrectly identifying it as IPv4... I can see in the code the logic for listen addresses is based just on the length of the parsed data structures, so something in the handling of the internal dict/list is giving the wrong length for the IP address data. Shot in the dark, but try using `0:0:0:0:0:0:0:0001` in the stubby.yml file....

I'll try to test this on 21.10 linux to see if is a Ubuntu version or a platform problem
@wtoorop do you have a raspberry pi available for testing?

**Comment by mynameisneo7** [comment]
Just noticed I hadn't given a version number in my original report, and turns out Ubuntu 21.10 has Stubby 0.3.0 (at least for arm?). So I grabbed the 0.4.0 stable tarball and built by hand and things work normally.

I was also momentarily puzzled by the same issue reported in #281 but despite the IPv6 port not being listed in netstat, it does still work normally via IPv6. With the sysctl change though, netstat reports it normally and functions normally of course.

Sorry if I wasted anyone's time! Though, not entirely sure why Ubuntu hasn't bumped their package yet.
